### ISSUE TITLE
Migrate away from google.com gcp project kubebuilder

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/k8s.io#2647]
Part of umbrella issue to migrate away from google.com gcp projects: https://github.com/kubernetes/k8s.io/issues/1469

We didn't notice this because it doesn't show up anywhere in kubernetes/test-infra, but apparently kubebuilder uses a google-internal project (ref: https://github.com/kubernetes/k8s.io/issues/1469#issuecomment-909658529)

We need help from someone from the subproject who has access to scope out what exactly is used and determine how best to migrate, e.g.
- maybe this can be satisfied by a k8s-staging-kubebuilder project?
- maybe the staging project needs some special-case functionality enabled?
- eg.

/wg k8s-infra
/sig api-machinery
/priority important-soon
/milestone v1.23

/assign @leilajal
I'm assigning to you to help with scoping out how the internal project is used and what needs to be moved over.  Feel free to reassign to someone who is more familiar with kubebuilder's release process

**Comment by camilamacedo86** [comment]
Hi folks, 

Following is what kubebuilder and controller-runtime use from google cloud currently:

a) We use it to build and provide the images https://console.cloud.google.com/gcr/images/kubebuilder/GLOBAL/kube-rbac-proxy

b) We use it to build and provide the kubebuilder-tools (which has the artefacts to be used within envtest from controller-runtime): https://storage.googleapis.com/kubebuilder-tools

So, if it needs to be changed, we need to align and see how that should be changed and etc. Please, feel free to reach out to us via slack. Note that all projects consume both, so no longer having those available would be a breaking change that impacts everybody.

The name of the project in the Google Cloud is kubebuilder

By the way, we need help accessing the Google Cloud because I had access to keep those maintained but the email used for it no longer exists, and I am unable to reach out @kevindelgado who helped us to change the email.

**Comment by kevindelgado** [comment]
So sorry for the delay while I was on paternity leave. 

I just granted @camilamacedo86  the same access she had before for the project and her new email. For future reference, my managers @fedebongio and @leilajal also have ownership of the kubebuilder GCP project, in case any one of us can't be reached.

**Comment by camilamacedo86** [comment]
I wanted to document the most recent status:

- We've begun updating the kube-rbac-proxy image. However, the only reason this image is being generated by us is because the project isn't under a sig (https://github.com/kubernetes/test-infra/pull/29351/files and https://github.com/kubernetes-sigs/kubebuilder/pull/3362). Ideally, we shouldn't have to re-tag their image but rather consume the one they provide. We're awaiting the completion of its donation so that we can use their official image. More details can be found here: https://github.com/brancz/kube-rbac-proxy/issues/238. **Therefore, I would like to advogated in favor of kubebuilder no longer be responsable for re-tag the images and we deprecated/remove those logic asap the kube-rbac-proxy project be able to provide the images to us.** c/c @bihim

- After this step, we'll still require GCP to produce the binaries used by the env test (kubebuilder-tools). Further information can be found here: https://book.kubebuilder.io/reference/artifacts.html

From the current outlook, it seems unlikely that we will be able to avoid using GCP, unless there's a proposal to change the aforementioned binaries.

**Comment by camilamacedo86** [comment]
Just an update here:

We revisited this issue and identified two specific needs for continuing to use Google Cloud, as detailed in these discussions: [GitHub issue #2647 comment](https://github.com/kubernetes/k8s.io/issues/2647#issuecomment-1366836392) and [Slack message](https://kubernetes.slack.com/archives/CCK68P2Q2/p1711913605487359).

We attempted to proceed with migrating the kube-rbac-proxy image as per this link: [Google Cloud Registry link](https://console.cloud.google.com/gcr/images/kubebuilder/GLOBAL/kube-rbac-proxy). However, we encountered a hurdle due to the project not being under Kubernetes-SIG. We're actively working on resolving this.

**Comment by sbueringer** [comment]
We would look into what we can do about the binaries for envtest from the controller-runtime / controller-tools side.
I think it makes sense to move them over in some way to controller-runtime / controller-tools where setup-envtest already is.

**Comment by camilamacedo86** [comment]
It's crucial to emphasize that, should we find ourselves in a position to generate or promote artifacts elsewhere, it still really important we keep GCP for a long period running and promoting the artifcts that were generated in the past so that we do not broke the projects and give a good grace of time for projects be able to change.

**In the case of the `kube-rbac-proxy` if we stop to promote it in the current location it will be very critical and will broke a lot of projects in production that are using it.**

**Comment by camilamacedo86** [comment]
@upodroid 

This was closed and we cannot close this one. 
I think that was a mistake.


Just to clarify here: We did not migrated away from GCP yet and indeed after we migrate we need to keep things running their as much as possible and for a long period.

**Comment by BenTheElder** [comment]
It looks to me like so far:
- we have plans to sort out the binaries
- we are phasing out the kube-rbac-proxy image

Are there any other container images that should be considered? I see many other in gcr.io/kubebuilder ...

**Comment by camilamacedo86** [comment]
Hi @BenTheElder,

You are right. For the Kubebuilder project itself, those are still in use. Beyond that, the majority of what we have there is from older versions.

However, I took another look in the GCP and found that we are using the project to generate the image gcr.io/kubebuilder/pr-verifier:$TAG_NAME. This image, generated by the project https://github.com/kubernetes-sigs/kubebuilder-release-tools, is used in the CI to verify PR emojis for release notes.I think @vincepri and @sbueringer are who has been taking care of https://github.com/kubernetes-sigs/kubebuilder-release-tools.

So, we need a plan as well for this image since it will no longer be available after `March 18, 2025`.

**Comment by sbueringer** [comment]
"taking care" is relative :) But good point, didn't think about this one. I think there is no reason why it couldn't use the regula image promotion

**Comment by camilamacedo86** [comment]
Hi @sbueringer,

I think the plan with would use  the shared e2e infra and  the regular image promotion too. +1

**Comment by sbueringer** [comment]
Quick update on the setup-envtest / envtest binary situation.

PR merged to retrieve the envtest binaries from controller-tools releases: https://github.com/kubernetes-sigs/controller-runtime/pull/2811
Also cherry-pick to CR release-0.18: https://github.com/kubernetes-sigs/controller-runtime/pull/2837

This means:
* setup-envtest@release-0.18 will retrieve the binaries from GCS per default, but allows to download from controller-tools releases (I didn't want to change the default behavior on a release branch / patch release)
* setup-envtest@main/latest/ upcoming 0.19 (release ~ in August) will download from controller-tools releases per default, while still allowing from the GCS bucket

**Comment by camilamacedo86** [comment]
### Conclusion and Latest Status

The GCP project must be active until **March 18, 2025**. After that date then theoretically Kubebuilder GCP project can be shotdown, Google has announced that images under `gcr.io/kubebuilder` will no longer be available. While Kubebuilder no longer promotes new images or artifacts through this repository, some projects built with older versions are using images like `gcr.io/kubebuilder/kube-rbac-proxy` in production environments.

**Therefore, we must keep it running for as long as possible, until these older projects have fully migrated away. However, the goal of this issue was ensure that we migrate away from GCP project and from our side, everything has been done to discontinue the GCP usage.** 

### Detailed info about what was done to allow us move away: 

- **Kubebuilder releases** are now managed via GoReleaser and GitHub Actions. For over 1 or 2 years, they no longer rely on GCP to trigger builds.
- **images from kube-rbac-proxy project used to protect metrics endpoints** Projects initialized with Kubebuilder release versions `>= [v3.15.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v3.15.0)` are no longer scaffolded with the `kube-rbac-proxy` dependency. Instead, they use the built-in features provided by controller-runtime (`filters.WithAuthenticationAndAuthorization`) by default. You can see an example [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v4.2.0/testdata/project-v4/cmd/main.go#L115-L121). Furthermore, users are given an extra optional option to protect their metrics endpoint with network policies, as shown [here](https://github.com/kubernetes-sigs/kubebuilder/blob/c7cde5172dc8271267dbf2899e65ef6f9d30f91e/testdata/project-v4/config/default/kustomization.yaml#L30-L34). This change was communicated via the mailing list, Slack channels, release notes, and a discussion topic. For more details, see the discussion: https://github.com/kubernetes-sigs/kubebuilder/discussions/3907
- **kubebuilder-tools tarballs which provides the EnvTest binaries** The artifacts provided via https://storage.googleapis.com/kubebuilder-tools, which contain binaries to allow users use the Controller-Runtime ENVTEST library to test their controllers, are now built and released via controller-tools. You can find the releases:https://github.com/kubernetes-sigs/controller-tools/releases. Since the release of controller-runtime [v0.19](https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.19.0), the [setup-envtest](https://github.com/kubernetes-sigs/controller-runtime/tree/main/tools/setup-envtest) downloads tarballs from the new location. Any projects generated with Kubebuilder `>= [v4.2.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v4.2.0)`—which adds support for controller-runtime v0.19 and Kubernetes 1.31—will use this new location. Projects using Kubernetes `1.31` for testing will need to adopt the new location, which ensures a smooth transition. This was communicated via the mailing list, Slack channels, release notes, and a discussion topic. For further information, see the discussion: https://github.com/kubernetes-sigs/kubebuilder/discussions/4082.
- The image used in the GitHub Action to validate the PR title (from [kubebuilder-release-tools](https://github.com/kubernetes-sigs/kubebuilder-release-tools)) will also break once GCP images are unavailable. This project is not actively maintained, but Kubebuilder has implemented a simple solution to achieve the same functionality with GitHub Actions and shell scripts. The PR to update the documentation is [here](https://github.com/kubernetes-sigs/kubebuilder-release-tools/pull/61/files). Once this PR is merged, I am commit to share the update via the mailing list and channels to ensure everyone is informed.

**Therefore, from our side, everything is ready. We just need to give projects enough time to migrate before we fully discontinue the GCP usage. We can plan shot it down in April/2025, just ping me before more forward for we ensure that all is fine** 

c/c @BenTheElder @sbueringer @fedebongio

**Comment by camilamacedo86** [comment]
### Conclusion and Latest Status

The GCP project must be active until **March 18, 2025**. After that date then theoretically Kubebuilder GCP project can be shotdown, Google has announced that images under `gcr.io/kubebuilder` will no longer be available. While Kubebuilder no longer promotes new images or artifacts through this repository, some projects built with older versions are using images like `gcr.io/kubebuilder/kube-rbac-proxy` in production environments.

**Therefore, we must keep it running for as long as possible, until these older projects have fully migrated away. However, the goal of this issue was ensure that we migrate away from GCP project and from our side, everything has been done to discontinue the GCP usage.** 

> 🚀  **We can plan to shut it down in April 2025. Please ping me before moving forward so we can work together to ensure everything is in order.**


### Detailed info about what was done to allow us move away: 

- **Kubebuilder releases** are now managed via GoReleaser and GitHub Actions. For over 1 or 2 years, they no longer rely on GCP to trigger builds.
- **images from kube-rbac-proxy project used to protect metrics endpoints** Projects initialized with Kubebuilder release versions `>= [v3.15.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v3.15.0)` are no longer scaffolded with the `kube-rbac-proxy` dependency. Instead, they use the built-in features provided by controller-runtime (`filters.WithAuthenticationAndAuthorization`) by default. You can see an example [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v4.2.0/testdata/project-v4/cmd/main.go#L115-L121). Furthermore, users are given an extra optional option to protect their metrics endpoint with network policies, as shown [here](https://github.com/kubernetes-sigs/kubebuilder/blob/c7cde5172dc8271267dbf2899e65ef6f9d30f91e/testdata/project-v4/config/default/kustomization.yaml#L30-L34). This change was communicated via the mailing list, Slack channels, release notes, and a discussion topic. For more details, see the discussion: https://github.com/kubernetes-sigs/kubebuilder/discussions/3907
- **kubebuilder-tools tarballs which provides the EnvTest binaries** The artifacts provided via https://storage.googleapis.com/kubebuilder-tools, which contain binaries to allow users use the Controller-Runtime ENVTEST library to test their controllers, are now built and released via controller-tools. You can find the releases:https://github.com/kubernetes-sigs/controller-tools/releases. Since the release of controller-runtime [v0.19](https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.19.0), the [setup-envtest](https://github.com/kubernetes-sigs/controller-runtime/tree/main/tools/setup-envtest) downloads tarballs from the new location. Any projects generated with Kubebuilder `>= [v4.2.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v4.2.0)`—which adds support for controller-runtime v0.19 and Kubernetes 1.31—will use this new location. Projects using Kubernetes `1.31` for testing will need to adopt the new location, which ensures a smooth transition. This was communicated via the mailing list, Slack channels, release notes, and a discussion topic. For further information, see the discussion: https://github.com/kubernetes-sigs/kubebuilder/discussions/4082.
- **GitHub Action to validate the PR title (from [kubebuilder-release-tools](https://github.com/kubernetes-sigs/kubebuilder-release-tools))** `gcr.io/kubebuilder/pr-verifier:$TAG_NAME` will also break once GCP images are unavailable. This project is not actively maintained, but Kubebuilder has implemented a simple solution to achieve the same functionality with GitHub Actions and shell scripts. The PR to update the documentation is [here](https://github.com/kubernetes-sigs/kubebuilder-release-tools/pull/61/files). Once this PR is merged, I am commit to share the update via the mailing list and channels to ensure everyone is informed.

**🎉 Huge thanks to everyone who helped and contributed to this significant effort across many projects!
It would never have been possible without everyone's help.**


c/c @BenTheElder @sbueringer @fedebongio

**Comment by camilamacedo86** [comment]
### Request to Extend Availability of `gcr.io/kubebuilder/kube-rbac-proxy` Image  

Hi @BenTheElder @fedebongio,  

I hope you're doing well!  

I'd like your help in keeping the [`gcr. io/kubebuilder/kube-rbac-proxy`](gcr.io/kubebuilder/kube-rbac-proxy) image available for as long as possible. This image is critical, and its unavailability could disrupt many projects that rely on it, particularly those built with older versions of Kubebuilder or Operator SDK.  

To give an idea of how many projects will be affected, see the result of the following GitHub search: [Search results for `gcr.io/kubebuilder/kube-rbac-proxy`](https://github.com/search?q=gcr.io%2Fkubebuilder%2Fkube-rbac-proxy&type=code).  

---

### Key Points:  

**Kubebuilder no longer uses this image**. We have communicated this change and have been working on providing helpers for users to protect their metrics endpoint in other ways.  
- By default, we replaced its functionality with a feature provided by Controller Runtime: [`WithAuthenticationAndAuthorization`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/metrics/filters#WithAuthenticationAndAuthorization).  
- On February 16 was when we were reached out and then **any project built with Kubebuilder releases after May 15 no longer creates this dependency**.  [Announcement](https://github.com/kubernetes-sigs/kubebuilder/discussions/3907)
- **All other artifacts from GCP have been transitioned and are either no longer in use or not critical.**  
- **No new images have been promoted** (to encourage people to move forward, address CVEs, etc.).  

---

### Problem Statement:  

Any solution still relying on this image will fail to install or upgrade. **Pods will not run if they cannot pull the image.**  

I'd like to stress how crucial it is to avoid breaking existing setups. @DirectXMan12's original intent behind re-building and re-tagging this image was to ensure its long-term availability and prevent disruptions.  

---

### Proposed Solution:  

Could we consider **moving this image to Artifact Registry** and extending its availability?  

![Screenshot 2024-11-20 at 19 47 21](https://github.com/user-attachments/assets/188ddf44-46f7-4c22-9165-737c88106c6f)

Is that something that would be possible to consider and get help with? 

---

### Actions From My Side (by the End of 2024):  

1. I will create a **clear guidance document and communication plan** to assist users in replacing the image, including necessary code changes and decision-making points.  
2. I am considering developing a **shell script to identify repositories on GitHub** that still use this image, automatically open issues to alert maintainers, and provide them with this guidance.  
3. I will push **additional communications to further encourage users to transition away** from this dependency.  

---

Thank you, and I look forward to hearing your thoughts!  

Best regards,  
Camila Macedo

**Comment by upodroid** [comment]
> * To mitigate the risks we manually mirror the images to [gcr.io/k8s-staging-kubebuilder/kube-rbac-proxy](https://explore.ggcr.dev/?repo=gcr.io%2Fk8s-staging-kubebuilder%2Fkube-rbac-proxy)

This should be `registry.k8s.io/kubebuilder/kube-rbac-proxy`

**Comment by camilamacedo86** [comment]
To raise awareness within the community, I created a script to identify projects still using the legacy image and to open issues in their repositories. However, I want to acknowledge a few limitations:

- The GitHub API has usage restrictions that I attempted to work around, but I cannot guarantee that the script identifies all public projects.
- Not all projects are public, nor can we assume that all are hosted on GitHub.

Despite these constraints, I hope this initiative will bring significant awareness and help many developers transition away from the legacy image.

**Comment by camilamacedo86** [comment]
Hi @ameukam,  

Thanks for the question! I think we are looking to help. That is amazing 🚀  

The issue isn’t with Kubebuilder itself. Kubebuilder doesn’t use the GCP project anymore. **However, thousands of projects worldwide still depend on [gcr.io/kubebuilder/kube-rbac-proxy](https://console.cloud.google.com/artifacts/docker/kubebuilder/us/gcr.io/kube-rbac-proxy) ([example](https://github.com/search?q=gcr.io%2Fkubebuilder%2Fkube-rbac-proxy&type=code)). Developers need to update their projects, release new versions, and users must stop using the old versions of those solutions, which is a significant effort worldwide.**  

### What would help the most?  

> **If we can ensure that people can still `pull gcr.io/kubebuilder/kube-rbac-proxy` for as long as possible** 

It will greatly help mitigate the impact. The more time we can give developers to update their solutions and users to upgrade their environments, the better.  **Here are two potential approaches that I thought about:**  

- **Create a redirect from `gcr.io/kubebuilder/kube-rbac-proxy` to `registry.k8s.io/kubebuilder/kube-rbac-proxy`**: Note that we manually mirror the legacy images to an alternative location: https://explore.ggcr.dev/?repo=registry.k8s.io%2Fkubebuilder%2Fkube-rbac-proxy
- **Migrate from GCR to AR (assuming that means not changing the domain) and do our best to keep things running as much as we can**: This would allow users to pull the image still using the same address: `docker pull gcr.io/kubebuilder/kube-rbac-proxy` for as long as possible

🙌 **Your support here would make a big difference. Thank you!**  

---

> ## TL;DR Context  
> 
> [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) is a framework for building cloud-native solutions. Tools like [Operator-SDK](https://github.com/operator-framework/operator-sdk), which serve a similar purpose, share the same foundation. Historically, the [`kube-rbac-proxy`](https://github.com/brancz/kube-rbac-proxy) image was scaffolded into projects to secure the metrics endpoint.  Therefore, any project built in the past with those tools and that did not upgrade its scaffold uses the impacted image.
> 
> Before I joined the project, a former project admin made the decision to adopt this solution and began to re-tag and publish the image under the GCP project to avoid ecosystem outages. **While I don’t agree with the decision to add this dependency or take responsibility for publishing it, the intention was to prevent exactly the kind of disruption we are facing now. :-(**

**Comment by BenTheElder** [comment]
SIG K8s Infra and this repo don't own gcr.io/kubebuilder or track those changes, however we do need to track the Kubernetes project moving to community infra.

It sounds like we're done moving community infra in OSS, and it is just downstreams that need to migrate, with is not SIG K8s Infra's responsibility.

**Comment by camilamacedo86** [comment]
Hi @BenTheElder,

I understand that this issue was previously opened to track the need to migrate Kubebuilder away from GCP. Could we please keep it open until the final day that the GCP project will be operational?

I prefer not to have it closed because:

1. a) As we discussed and highlighted above, we need to make our best effort to keep it running until at least March 18, 2025, or longer if possible, due to the impact caused by the usage of the image gcr.io/kubebuilder/kube-rbac-proxy. 

2. b) **I'm very concerned that if someone sees it closed, they might assume it's acceptable to decommission GCP**. Ideally, we want to keep the image gcr.io/kubebuilder/kube-rbac-proxy pullable for as long as possible. 

Thank you for your attention to this matter.

**Comment by BenTheElder** [comment]
> Migrate from GCR to AR (assuming that means not changing the domain) and do our best to keep things running as much as we can: This would allow users to pull the image still using the same address: docker pull gcr.io/kubebuilder/kube-rbac-proxy for as long as possible

This was done already, though we don't intend to advertise widely specifics about this, only that users should migrate to registry.k8s.io or their own [mirror](https://github.com/kubernetes/registry.k8s.io/tree/main/docs/mirroring).


> I understand that this issue was previously opened to track the need to migrate Kubebuilder away from GCP. Could we please keep it open until the final day that the GCP project will be operational?

This is not an appropriate place to track, it is not owned by Kubernetes and SIG K8s Infra is not responsible for it (this is the SIG K8s Infra issue tracker).

This is part of a larger set of tracking issues for ensuring that _The Kubernetes project_ has successfully transitioned our infrastructure. We have done that here.

> a) As we discussed and highlighted above, we need to make our best effort to keep it running until at least March 18, 2025, or longer if possible, due to the impact caused by the usage of the image gcr.io/kubebuilder/kube-rbac-proxy.

That is happening independent of this issue, these issues are being tracked for SIG K8s Infra, who does not control this project. This issue is part of the overall migration tracking effort and the migration is complete. 

> b) I'm very concerned that if someone sees it closed, they might assume it's acceptable to decommission GCP. Ideally, we want to keep the image gcr.io/kubebuilder/kube-rbac-proxy pullable for as long as possible.

There is no one else at GCP looking at this aside from @fedebongio who understands.

No one else who doesn't understand is going to go looking for a github issue related to this project.

**Comment by BenTheElder** [comment]
> There is no one else at GCP looking at this aside from @fedebongio who understands.

Er, not quite true, we had help from a member of the artifact registry team who handled the migration for us 🙏, but only with temporary access, they're not going to shut it down and don't plan further involvement.

**Comment by camilamacedo86** [comment]
Hi @BenTheElder 

Thank you a lot for all your help, support and explanations
Either for the migration ❤️ . 
It makes sense. 

> _Just to let you know:_
> 
> _I am continuing to work on leveraging the GitHub API to identify unique public projects and ensure they have issues raised in their repositories to help contributors move forward. For instance, you can refer to this [example issue](https://github.com/validator-labs/validator-plugin-network/issues/297). Regarding KubeBuilder, its latest scaffolds no longer rely on any external registry or additional image dependencies, so this particular problem should not occur in the future._

Again, thank you a lot for all your support 🙌
